src/relay/pass/type_infer.cc

```diff
@@ -219,14 +194,63 @@ class TypeInferencer : private ExprFunctor<Type(const Expr&)> {
           EnvFunc::Get("tvm.relay.type_relation.TupleGetItem").node_);
     }
     Type tuple_type = GetType(op->tuple);
-    Type rtype = IncompleteTypeNode::make(TypeVarNode::Kind::kType);
+    Type rtype = IncompleteTypeNode::make(Kind::kType);
     auto attrs = make_node<TupleGetItemAttrs>();
     attrs->index = op->index;
     solver_.AddConstraint(TypeRelationNode::make(
         tuple_getitem_rel_, {tuple_type, rtype}, 1, Attrs(attrs)), GetRef<TupleGetItem>(op));
     return rtype;
   }
 
+  void VisitPattern_(const PatternConstructorNode* con, const Type& t) {
+    CHECK(mod_.defined())
+      << "Cannot do type inference without a environment:"
+      << con->constructor->name_hint;
+    TypeData td = mod_->type_definitions.at(con->constructor->belong_to);
+
+    // we can expect a certain number of arguments
+    Array<Type> unknown_args;
+    for (size_t i = 0; i < td->tv.size(); i++) {
+      unknown_args.push_back(IncompleteTypeNode::make(Kind::kType));
+    }
+    Type expected = TypeCallNode::make(con->constructor->belong_to, unknown_args);
+    Type unified = Unify(t, expected, GetRef<NodeRef>(con));
+
+    auto* tc = unified.as<TypeCallNode>();
+    CHECK(tc) << "must be type call";
```

Comment:
I am not sure but should we use error reporter there?

Reply:
Yeah, I'll change it